crates/common/tedge_config_macros/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+            #[tedge_config(example = "/etc/tedge/device-certs/tedge-certificate.pem")]
+            #[doku(as = "PathBuf")]
+            #[serde(alias = "certfile")]
+            cert_file: Utf8PathBuf,
+
+            /// Path to the key file which is used by the external MQTT listener
+            #[tedge_config(note = "This setting shall be used together with `mqtt.external.cert_file` for external connections.")]
+            #[tedge_config(example = "/etc/tedge/device-certs/tedge-private-key.pem")]
+            #[doku(as = "PathBuf")]
+            #[serde(alias = "keyfile")]
+            key_file: Utf8PathBuf,
+        }
+    }
+}
+
+fn default_device_key(location: &TEdgeConfigLocation) -> Utf8PathBuf {
+    location
+        .tedge_config_root_path()
+        .join("device-certs")
+        .join("tedge-private-key.pem")
+}
+
+fn default_device_cert(location: &TEdgeConfigLocation) -> Utf8PathBuf {
+    location
+        .tedge_config_root_path()
+        .join("device-certs")
+        .join("tedge-certificate.pem")
+}
+
+fn default_mqtt_port() -> NonZeroU16 {
+    NonZeroU16::try_from(1883).unwrap()
+}
+
+#[test]
+fn root_cert_path_default() {
+    let dto = TEdgeConfigDto::default();
+    let reader = TEdgeConfigReader::from_dto(&dto, &TEdgeConfigLocation);
+    assert_eq!(reader.az.root_cert_path, "/etc/ssl/certs");
+}
+
+#[test]
+fn writable_keys_can_be_parsed_from_aliases() {
+    let _: WritableKey = "az.mapper.timestamp".parse().unwrap();
+    let _: WritableKey = "azure.mapper.timestamp".parse().unwrap();
+}
+
+#[test]
+fn readable_keys_can_be_parsed_from_aliases() {
+    let _: ReadableKey = "az.mapper.timestamp".parse().unwrap();
+    let _: ReadableKey = "azure.mapper.timestamp".parse().unwrap();
+}
+
+#[test]
+fn default_from_path_uses_the_correct_default() {
+    #![allow(unused_variables)]
+    define_tedge_config! {
+        test: {
+            #[tedge_config(default(value = "DEFAULT_VALUE_FOR_ONE"))]
+            one: String,
+            #[tedge_config(default(from_path = "test.one"))]
+            two: String,
+        }
+    }
+    let dto = TEdgeConfigDto::default();
+    assert_eq!(
+        TEdgeConfigReader::from_dto(&dto, &TEdgeConfigLocation)
+            .test
+            .two,
+        "DEFAULT_VALUE_FOR_ONE"
+    );
+}
+
+#[test]
+fn default_from_path_uses_the_value_of_other_field_if_set() {
+    #![allow(unused_variables)]
+    define_tedge_config! {
+        test: {
+            #[tedge_config(default(value = "DEFAULT_VALUE_FOR_ONE"))]
+            one: String,
+            #[tedge_config(default(from_path = "test.one"))]
+            two: String,
+        }
+    }
+    let mut dto = TEdgeConfigDto::default();
+    dto.test.one = Some("UPDATED_VALUE".into());
+    assert_eq!(
+        TEdgeConfigReader::from_dto(&dto, &TEdgeConfigLocation)
+            .test
+            .two,
+        "UPDATED_VALUE"
+    );
+}
+
+#[test]
+fn default_from_path_uses_its_own_value_if_both_are_set() {
```

Comment:
To improve test coverage, it would be good to add a test where `test.two` have been updated but `test.one` is still the default.